app/src/main/java/sushi/hardcore/droidfs/explorers/BaseExplorerActivity.kt

```diff
@@ -408,22 +408,99 @@ open class BaseExplorerActivity : BaseActivity() {
                 items.clear()
                 break
             } else {
-                items.add(OperationFile.fromExplorerElement(ExplorerElement(fileName, 1, -1, -1, currentDirectoryPath)))
+                items.add(OperationFile.fromExplorerElement(ExplorerElement(fileName, 1, -1, -1, folderPath)))
             }
         }
         if (items.size > 0) {
-            checkPathOverwrite(items, currentDirectoryPath) { checkedItems ->
+            checkPathOverwrite(items, folderPath) { checkedItems ->
                 checkedItems?.let {
-                    fileOperationService.importFilesFromUris(checkedItems, uris){ failedItem ->
+                    fileOperationService.importFilesFromUris(checkedItems, uris){ failedReason ->
                         runOnUiThread {
-                            callback(failedItem)
+                            callback(failedReason)
                         }
                     }
                 }
             }
         }
     }
 
+    protected fun importFilesFromUris(uris: List<Uri>, callback: (String?) -> Unit) {
+        importFilesFromUrisIntoFolder(uris, currentDirectoryPath, callback)
+    }
+
+    private fun importDirectory(rootFile: DocumentFile, parentPath: String): String? {
+        val selfPath = PathUtils.pathJoin(parentPath, rootFile.name!!)
+        if (rootFile.isFile) {
+            return null
+        }
+
+        // List immediate children
+        val childFiles = arrayListOf<Uri>()
+        val childFolders = arrayListOf<DocumentFile>()
+        val childrenFetcher = Thread {
+            val children = rootFile.listFiles()
+            childFiles.addAll(children.filter { f -> f.isFile }.map { f -> f.uri })
+            childFolders.addAll(children.filter { f -> f.isDirectory })
+        }
+        childrenFetcher.start()
+        childrenFetcher.join()
```

Comment:
Why do you spawn a thread then immediately join it ? Why don't you just do the mapping on the current thread ?

Reply:
It turns out that `listFiles`, `isFile` and `isDirectory` are not interruptible. And with a folder of ~400 files, this took one minute on my Pixel 4a.  

So if a user imports a big folder, and decides to cancel it, they might have to wait for a whole minute just for cancellation to complete. (This can be tested by removing the thread, and using a folder that contains two folders, each with ~400 files in them)

And the simplest way to fix that, was to put the non-interruptible calculations on a thread, and then wait for that thread, because `join` is interruptible.